### PR TITLE
[somneo] Add alarm support and other improvements

### PR DESCRIPTION
**Pull request by 0x4d4d**
# Description

- Add support to control all the alarm clocks of the device
- Fix a issue that some channels used the wrong State class

# Testing

[org.openhab.binding.somneo-3.4.4-SNAPSHOT.jar](https://github.com/0x4d4d/openhab-addons/releases/download/somneo-alarm/org.openhab.binding.somneo-3.4.4-SNAPSHOT.jar)
[org.openhab.binding.somneo-4.0.0-SNAPSHOT.jar](https://github.com/0x4d4d/openhab-addons/releases/download/somneo-alarm/org.openhab.binding.somneo-4.0.0-SNAPSHOT.jar)

### CODE REVIEW

**Review by clinique — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.binding.somneo/src/main/java/org/openhab/binding/somneo/internal/SomneoBindingConstants.java
Comment (0x4d4d) [reply bundles/org.openhab.binding.somneo/src/main/java/org/openhab/binding/somneo/internal/SomneoBindingConstants.java:61]: Thank you for the review. What exactly do you mean by separate? Rename the group id alarm because it is too similar to the other alarm groups ids with an index (alarm{index})?

File: bundles/org.openhab.binding.somneo/src/main/java/org/openhab/binding/somneo/internal/SomneoBindingConstants.java
Comment (0x4d4d) [reply bundles/org.openhab.binding.somneo/src/main/java/org/openhab/binding/somneo/internal/SomneoBindingConstants.java:75]: The regex is mainly needed to extract the alarm index from the group id.

File: bundles/org.openhab.binding.somneo/src/main/java/org/openhab/binding/somneo/internal/SomneoHandler.java
Comment (jlaur) [inline bundles/org.openhab.binding.somneo/src/main/java/org/openhab/binding/somneo/internal/SomneoHandler.java:532]: There is no need to check if cancelled.

File: bundles/org.openhab.binding.somneo/src/main/resources/OH-INF/i18n/somneo_de.properties
Comment (jlaur) [inline bundles/org.openhab.binding.somneo/src/main/resources/OH-INF/i18n/somneo_de.properties:1]: This file must be omitted. Translations should be provided through Crowdin after the PR is merged. See https://www.openhab.org/docs/developer/utils/i18n.html#managing-translations

File: bundles/org.openhab.binding.somneo/src/main/resources/OH-INF/thing/thing-types.xml
Comment (jlaur) [inline bundles/org.openhab.binding.somneo/src/main/resources/OH-INF/thing/thing-types.xml:168]: Would you like to use the label and description of the channel type now or still overwrite it here?

Also, since the channel type changed, you need to provide update instructions for the channel. See https://next.openhab.org/docs/developer/bindings/thing-xml.html#updating-thing-types

Update instructions also apply to new channels added to existing things.

File: bundles/org.openhab.binding.somneo/src/main/resources/OH-INF/thing/thing-types.xml
Comment (jlaur) [inline bundles/org.openhab.binding.somneo/src/main/resources/OH-INF/thing/thing-types.xml:250]:
```suggestion
		<label>Alarm Clock</label>
```

File: bundles/org.openhab.binding.somneo/src/main/java/org/openhab/binding/somneo/internal/SomneoBindingConstants.java
Comment (jlaur) [inline bundles/org.openhab.binding.somneo/src/main/java/org/openhab/binding/somneo/internal/SomneoBindingConstants.java:65]:
```suggestion
    public static final String CHANNEL_ALARM_REPEAT_DAY = "alarm%d#repeatDay";
```

**Review by jlaur — CHANGES_REQUESTED** [review]
Thanks for the improvements! 👍 The code looks good, I have added only some minor comments.

File: bundles/org.openhab.binding.somneo/src/main/java/org/openhab/binding/somneo/internal/SomneoBindingConstants.java
Comment (jlaur) [reply bundles/org.openhab.binding.somneo/src/main/java/org/openhab/binding/somneo/internal/SomneoBindingConstants.java:61]: @clinique - can you help clarifying this?

File: bundles/org.openhab.binding.somneo/src/main/resources/OH-INF/thing/thing-types.xml
Comment (0x4d4d) [reply bundles/org.openhab.binding.somneo/src/main/resources/OH-INF/thing/thing-types.xml:200]: This switch indicates if this alarm is set up or configured. This does not mean that the alarm is active. On API view the boolean says if the alarm on this index has an alarm or if they are only default values.
What would you prefer set or configured?

Comment (jlaur) [reply bundles/org.openhab.binding.somneo/src/main/resources/OH-INF/thing/thing-types.xml:200]: In that case I have no better suggestion, you can keep it as is. I was somehow confused if it meant "set".

File: bundles/org.openhab.binding.somneo/src/main/resources/OH-INF/thing/thing-types.xml
Comment (0x4d4d) [reply bundles/org.openhab.binding.somneo/src/main/resources/OH-INF/thing/thing-types.xml:33]: Yes, the device API provides a static list of 16 alarms. The first two are always present and cannot be reset (they do not have the configured switch). The others can be reset to the default values.
For comparison the mobile app from Philips works like the standard alarm clock app. Up to 16 alarms can be added and all except the first two can be deleted. Unfortunately, I could not map this user experience in OpenHAB.

Comment (jlaur) [reply bundles/org.openhab.binding.somneo/src/main/resources/OH-INF/thing/thing-types.xml:33]: Can you paste a screenshot how these channel groups like in the UI? Would save me some time trying to understand. 🙂

Comment (0x4d4d) [reply bundles/org.openhab.binding.somneo/src/main/resources/OH-INF/thing/thing-types.xml:33]: image

Comment (jlaur) [reply bundles/org.openhab.binding.somneo/src/main/resources/OH-INF/thing/thing-types.xml:33]: Thanks! I think it confirms my suspicion for posting the comment: Shouldn't there be individual labels for the channel groups so that they can be distinguished not only by channel UIDs? I.e. "Wecker 2", "Wecker 3" etc. in the group titles above.

File: bundles/org.openhab.binding.somneo/src/main/java/org/openhab/binding/somneo/internal/SomneoBindingConstants.java
Comment (jlaur) [reply bundles/org.openhab.binding.somneo/src/main/java/org/openhab/binding/somneo/internal/SomneoBindingConstants.java:61]: @clinique - gentle ping.

Comment (clinique) [reply bundles/org.openhab.binding.somneo/src/main/java/org/openhab/binding/somneo/internal/SomneoBindingConstants.java:61]: `alarm` is the group name, while `snooze` is the channel itself. You could directly target it using `channelUID.getIdWithoutGroup();` in your handler.

Comment (jlaur) [reply bundles/org.openhab.binding.somneo/src/main/java/org/openhab/binding/somneo/internal/SomneoBindingConstants.java:61]: @0x4d4d - did you see @clinique's reply?

Comment (0x4d4d) [reply bundles/org.openhab.binding.somneo/src/main/java/org/openhab/binding/somneo/internal/SomneoBindingConstants.java:61]: @jlaur Sorry, I had commented but it had the pending label and don't saw it.

Comment (jlaur) [reply bundles/org.openhab.binding.somneo/src/main/java/org/openhab/binding/somneo/internal/SomneoBindingConstants.java:61]: I don't see any issues with the current approach that should block this PR longer.

@clinique - considering @0x4d4d's latest comment, please continue and conclude the discussion here and do a follow-up PR if there is anything that can be improved without introducing unnecessary complexity or redundancy. I have seen similar comments from you in other PR's, so I think it would be nice to reach a conclusion/consensus.

You are also welcome to have a look at #14376 where I have something similar and couldn't find a better solution:

https://github.com/openhab/openhab-addons/pull/14376/files#diff-002ecf5a3879dda71cf4fcb82dbfef3f550c0f1655cfb16134fa3e4f3b431d69R42-R54

**Review by jlaur — APPROVED** [review]
LGTM

**Pull request merged**
